### ISSUE TITLE
`indexing_slicing` should not fire if a valid array index comes from a constant function that is evaluated at compile-time

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust-clippy#8348]
1. From a inline constant block (https://github.com/rust-lang/rust-clippy/issues/8348#issuecomment-1021303011).

```rust
#![feature(inline_const)]
#![forbid(clippy::indexing_slicing)]

const FOO: [i32; 1] = [1];

const fn idx() -> usize {
    0
}

pub fn main() {
    assert_eq!(FOO[const { idx() }], 1);
}
```

2. From a `const` declaration (https://github.com/rust-lang/rust-clippy/issues/8348#issuecomment-1021359538).

```rust
#![forbid(clippy::indexing_slicing)]

const FOO: [i32; 1] = [1];
const REF: &i32 = &FOO[idx()];

const fn idx() -> usize {
    0
}

pub fn main() {
  assert_eq!(REF, &1);
}
```

**Comment by xFrednet** [comment]
I think here we should distinguish if the type is an array with a fixed size, as that will be caught by the compiler or a slice. The lint should still trigger in the later case, as the range check will be done at runtime. This should be a simple type check :upside_down_face: 

@rustbot label +good-first-issue

**Comment by camsteffen** [comment]
I think this is functioning as designed. `const` functions are not evaluated at compile-time unless it is in a const context. If you change `idx` to return `1`, it will compile without warning and panic at runtime. Note that this is a restriction lint so it's generally not recommended unless you have a specific reason to use it.

**Comment by PatchMixolydic** [comment]
Here's [a disassembly of a similar code snippet showing that the call to `idx` isn't necessarily evaluated at compile time](https://godbolt.org/z/Eehefshf7). Incidentally, this lint still trips when the compiler is forced to evaluate `idx()` at compile time using an inline const block ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=153f3bf10c11494a0f28c3e71c813f28)):
```rust
#![feature(inline_const)]
#![forbid(clippy::indexing_slicing)]

const FOO: [i32; 1] = [1];

const fn idx() -> usize {
    0
}

pub fn main() {
    assert_eq!(FOO[const { idx() }], 1);
}
```

```
error: indexing may panic
  --> src/main.rs:11:16
   |
11 |     assert_eq!(FOO[const { idx() }], 1);
   |                ^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> src/main.rs:2:11
   |
2  | #![forbid(clippy::indexing_slicing)]
   |           ^^^^^^^^^^^^^^^^^^^^^^^^
   = help: consider using `.get(n)` or `.get_mut(n)` instead
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#indexing_slicing
```

This should presumably be allowed for the same reason as normal array indexing (any misuse would be caught by `const_err`).

**Comment by c410-f3r** [comment]
> `const` functions are not evaluated at compile-time unless it is in a const context

> Here's a disassembly of a similar code snippet showing that the call to idx isn't necessarily evaluated at compile time.

Oh, I forgot about this feature. Thanks for pointing out such behavior.

In addition to what @ PatchMixolydic said, the following snippet should also not fire

```rust
#![forbid(clippy::indexing_slicing)]

const FOO: [i32; 1] = [1];
const REF: &i32 = &FOO[idx()];

const fn idx() -> usize {
    0
}

pub fn main() {
  assert_eq!(REF, &1);
}
```

**Comment by c410-f3r** [comment]
Thanks @pitaj
